### DISCUSSION TITLE
How expensive is writing data into many small partitions in parquet format?

### DISCUSSION THREAD

**Discussion by mustafahasankhan** [duckdb/duckdb#9374] [Q&A]
I've been using the partition_by clause to output data into parquet files based on the YEAR, MONTH, and DAY columns. My goal is to capitalize on filter pushdown support via Hive partitioning when I read these parquet files later on.

The DuckDB documentation states:
`Writing data into many small partitions is expensive. It is generally recommended to have at least 100MB of data per partition.
`
To provide some context, I conducted a test with 1.1 million records, which yielded a directory structure resembling:

data
├── year=2021
│   ├── month=1
│   │   ├── day=10
│   │   │   ├── uuid.parquet

Each resulting parquet file was only a few hundred KBs, containing several thousand records. Initially, I allocated 10GB for the XMX value, but encountered an OOM error. After increasing the allocation to 20GB, I observed a significant memory spike towards the end, as illustrated here:
![image](https://github.com/duckdb/duckdb/assets/65130881/16ffbfa1-4e19-4e50-bbe6-e1f612bb2857)

Thus my question is, that whether are these small partitions causing high memory usage? 
Is it normal?

This is my sample code:
            `conn.createStatement().execute("COPY (SELECT * FROM raw_table) TO '/tmp/output' (FORMAT 'parquet', PARTITION_BY (YEAR, MONTH, DAY), OVERWRITE_OR_IGNORE, FILENAME_PATTERN '{uuid}')");
`

**Comment by l1t1** [comment]
how many partitions do you have? the key of the question isn't `small`, that is `many`.

**Reply by mustafahasankhan** [reply]
300+

**Comment by l1t1** [comment]
I run a small test
```
.timer on
create or replace table orders(m int,v int,j int);
insert into orders select i%12+1,i,j from range(360)t(i),range(1000)s(j);
copy (select 2000+(v//12)y,m,v,j from orders) TO 'orders_m' (FORMAT PARQUET, PARTITION_BY (m));
copy (select 2000+(v//12)y,m,v,j from orders) TO 'orders_y' (FORMAT PARQUET, PARTITION_BY (y));
copy (select 2000+(v//12)y,m,v,j from orders) TO 'orders_ym' (FORMAT PARQUET, PARTITION_BY (y,m));
```
the 3rd raises an error
```sql
Run Time (s): real 0.995 user 1.123207 sys 0.405603
Error: IO Error: Could not create directory: 'orders_ym\y=2010\m=6'
```
